Upgrade Flask dependency to 2.0.2 and refactor applications to improve static type checks

zenmonkeykstop opened the pull request:
## Status

Ready for review (but test plan TK)

## Description of Changes

Fixes #6154.

This PR updates the application Flask version to 2.0.2 and updates associated dependencies. It also refactors out the app.storage and app.instance_config dynamic attributes and corrects several typing errors, in order to allow for successul static typechecking via mypy.

- Dependency updates:
  - click from 6.7 to 8.0.3
  - flask-babel from 011.2 to 2.0.0
  - flask-wtf from 0.14.2 to 1.0.0
  - Flask from 1.0.2 to 2.0.2
  - itsdangerous from 0.24 to 2.0.1
  - jinja2 from 2.11.3 to 3.0.2
  - markupsafe from 1.1.1 to 2.0.1
  - redis from 3.3.6 to 3.5.3
  - rq from 1.1.0 to 1.10.0
  - werkzeug from 0.16.0 to 2.0.2

### Application refactor:
- Flask app.storage attribute has been removed in favour of a global Storage object, accessed via a `.get_default()` class method that creates the global object if it doesn't already exist, and then returns it.
- Flask app.instance_config attribute has been removed in favour of a global InstanceConfig object, accessed similarly to Storage.
- Numerous typing fixes have been made.
- datetime variables have been updated to be timezone-aware as required by Flask 2.0.2

### Test updates:
Tests have been updated to account for the refactor above. An app_storage fixture was added, returning a function-scoped Storage object. For tests which access storage directly, that fixture was added to their args and references like `current_app.storage.whatever()` were changed to `app_storage.whatever()`. For tests that access storage indirectly (for example, via a test journalist or source application),  the `Storage.get_default()` method must be mocked to return the `app_storage fixture`. This patch has been added in the `source_app` and `journalist_app` fixtures, so it will be applied automatically in tests that use them.

### General notes:
- There a few TODOs not addressed by this PR for various reasons
  - ~The `babel_instance` Flask app dynamic attribute has not been refactored out - it's added by `flask-babel` and an upstream change is probably required~ Fixed in #6220  
  - type checking is disabled for the CSRFError error handlers due to an issue similar to: https://github.com/pallets/flask/issues/4295
  - type checking is [disabled in the generic error handler ](https://github.com/freedomofpress/securedrop/pull/6217/commits/bae4effa98a9f9133fbb215ddbc0f003e97233c5#diff-24b8e3317d6c0ff33dfe29a1c3edfbb95a6ebfac51562e105f9214c55d1e955cR91) - error handling should be updated to remove the use of Flask's internal `app.error_handler_spec` structure
  - during initial testing, it was found that test requirements would override app requirements in the test Docker environment, meaning that updates to app requirements weren't always being applied. As a quick fix, the order in which requirements were installed was switched, so that app requirements would always be applied last. A better fix would be to clean up those requirements files to either keep them in sync or to remove application dependencies from the test requirements files altogether.

## Testing

### Storage-specific tests     
                                  
#### Secure temporary files    
                                                    
- check out this branch and run  `make dev-tor` (some latency will help when observing the tempfile behaviour)                          
- in another teminal, connect to the dev env  with `docker exec -t securedrop-dev-0 bash`                           
- in the dev env, monitor /tmp with `watch -n 1 ls -l /tmp`                                       
- connect to the Source Interface(SI) (you'll find the address listed in docker output just before the applications start) and upload a file greater than 512k.                                
  - [x] confirm that 2 files with a `.aes` extension are written and then deleted as the upload completes (the first is the gzipped initial submission, second is decompressed version)
- (optional) for bonus points, retain both files by adding a `delete=False` argument to the super().__init() in the SecureTemporaryFile class and submitting another file larger than 512k
  - [x] verify that both files contents "look" random. 
  - [x] verify that the size of the first file matches the gzipped size of the submission and the size of the second matches that of the original file.
                                                                                
#### Submissions and replies                                                       
- are files being saved to the right location?                                 
  - check out this branch and run  `make dev`                       
  - in another teminal, connect to the dev env  with `docker exec -t securedrop-dev-0 bash`       
  - in  the SI, start at least 2 source sessions, upload distinct files to both    
  - [x] verify that they were stored to the correct `/var/lib/securedrop/store/<UUID>/` paths for the respective sources
- are they being attributed to the correct source in the JI?                  
  - [x] in the Journalist Interface(JI), verify that the submissions are listed under the correct source page
- are replies only visible to the intended recipient in the SI?               
  - in the JI, reply to both sources                                                     
  - [x] in SI, confirm for each source that they can see their own reply only     
- are they encrypted?                
    - [x] in the JI, download the submissions and confirm that they can be decrypted with the submission key
- are they being deleted correctly (individually, en masse, or as part of a source wipe)
  - [x] delete one source's submissions and replies via the "files and messages" only option, confirm they are no longer present
  - [x] delete an individual submission from another source, confirm that the file was removed from the store
  - [x] delete the source in its entirety, confirm that the source store directory was deleted
- are source store directories being recreated correctly?   
  - in the SI, create another source and submit some docs/messages.                  
  - delete all files and messages from a source while preserving the source itself
  - in the dev env shell, manually delete the source's store directory                    
  - [x] attempt to submit a new doc as that source and confirm that the directory is recreated and the source submission is present
                                                                                
                                                                                
### Instance config tests                            
- check out this branch and run  `make dev`, then log in to the JI                                                 
 - on the **Admin > Instance Config** page, update the instance name.
   - [x] verify that the name is saved successfully and that the JI titles change to reflect it.
   - [x] load the SI and verify that the new name is used there as well.
 - on the **Admin > Instance Config** page, disable file uploads.
   - [x] load the SI and verify that sources only have the option to submit messages, not files.
                                                                                
### End-to-end tests                                                            
-  Perform as much of the [Acceptance Tests section ](https://github.com/freedomofpress/securedrop/wiki/2.1.0-Test-Plan#application-acceptance-testing)of the `2.1.0` test plan as you have time and patience for.
    - [x] verify that acceptance tests pass.
                                                                                
### Upgrade scenario         
- Follow the [upgrade scenario docs](https://docs.securedrop.org/en/stable/development/upgrade_testing.html#upgrade-testing-using-locally-built-packages)  to verify an upgrade from 2.1.0 to locally-built packages.      
  - [ ] verify that the upgrade scenario completes successfully       
  - [ ] verify that basic functionality is available after the upgrade.                                
  
## Deployment

These application changes will be deployed with the next release containing them. There are no schema changes or data migrations, but 2 of the alembic migration scripts were modified to reference the new global Storage object - as such, attention should be paid during testing to QA scenarios involving upgrades to verify that they work as expected.


## Checklist

### If you made changes to the server application code:

- [x] Linting (`make lint`) and tests (`make test`) pass in the development container

### If you made changes to the system configuration:

- [x] [Configuration tests](https://docs.securedrop.org/en/latest/development/testing_configuration_tests.html) pass

### If you made non-trivial code changes:

- [ ] I have written a test plan and validated it for this PR **TK**

Choose one of the following:

- [ ] I have opened a PR in the [docs repo](https://github.com/freedomofpress/securedrop-docs) for these changes, or will do so later
- [ ] I would appreciate help with the documentation
- [x] These changes do not require documentation

### If you added or updated a production code dependency:

Production code dependencies are defined in:

- `admin/requirements.in`
- `admin/requirements-ansible.in`
- `securedrop/requirements/python3/securedrop-app-code-requirements.in`

If you changed another `requirements.in` file that applies only to development
or testing environments, then no diff review is required, and you can skip
(remove) this section.

Choose one of the following:

- [ ] I have performed a diff review and pasted the contents to [the packaging wiki](https://github.com/freedomofpress/securedrop-debian-packaging/wiki)
- [x] I would like someone else to do the diff review - given the scope, looking for volunteers. So far we have:
  - [x] click from 6.7 to 8.0.3 - @conorsch
  - [x] flask-babel from 011.2 to 2.0.0 - @cfm
  - [x] flask-wtf from 0.14.2 to 1.0.0 - @zenmonkeykstop
  - [x] Flask from 1.0.2 to 2.0.2 - @zenmonkeykstop
  - [x] itsdangerous from 0.24 to 2.0.1 - @conorsch
  - [x] jinja2 from 2.11.3 to 3.0.2 - @legoktm
  - [x] redis from 3.3.6 to 3.5.3 - @legoktm
  - [x] rq from 1.1.0 to 1.10.0 - @eaon

zenmonkeykstop committed:
Replaced naive datetime.utcnow() calls with tz-aware equivalent.

Updated tests to use timezone-aware datetimes.

zenmonkeykstop committed:
Disabled pylint:E0237 false positives for g attributes

See https://github.com/pallets/flask/issues/4020 for more info
on this error.

zenmonkeykstop committed:
Replaced flask app.instance_config attribute with global InstanceConfig object

zenmonkeykstop committed:
Replaced Flask app.storage attribute with global Storage object

 - added global Storage object
 - updated source app to use global storage
 - updated journalist app to use global storage
 - updated loaddata.py to use global storage
 - updated shredder to use global storage
 - updated alembic migration scripts to use global storage object

zenmonkeykstop committed:
Removed Storage.get_default() refs from db model and store utility functions.

In models.py, the Submission and Reply constructors reference the Storage object to calculate file size information. Originally they used Storage.get_default() directly, but passing the Storage object as a parameter instead more cleanly separates the model from the application and simplifies the tests.

Similar considerations apply to the async_add_checksum_for_file() function in store.py.

legoktm commented on securedrop/journalist_app/__init__.py:
```diff
@@ -110,14 +110,10 @@ def _handle_http_exception(
 
-    @app.before_request
-    def load_instance_config() -> None:
-        app.instance_config = InstanceConfig.get_current()
```
If I'm understanding this correctly, this is a behavior shift now in that InstanceConfig is just initialized once at startup rather than on each request. Is that intentional? I haven't figured out how that works with the "valid_until" field yet...

zenmonkeykstop replied:
Actually the instance config is now retrieved every time InstanceConfig.get_default() is called, so it's  actually hitting the DB as much as before (or potentially more often if a single request has multiple .get_default()s) This reduces the benefit of a global object, so there's probably some opportunity for optimization there. The trick would be to detect configuration changes when they're made in the journalist interface, as we want those to be reflected immediately in the source app.

zenmonkeykstop replied:
Updated the InstanceConfig get_default() method to be lazier - it will now only hit the db if explicitly requested or if the global is `None`. With some tweaks to `@app.before_request`-decorate functions, changes should show up immediately in both apps without extra DB overhead.

Also, re `valid_until` - it might be a bit counter-intuitive. It records the historical info about previous configs, and when they were invalidated. (TBH I'm not sure we need to store said previous configs, but it does provide an audit trail for instance_config changes.)

legoktm commented on securedrop/store.py:
```diff
@@ -314,9 +329,12 @@ def save_file_submission(self,
+                             filename: typing.Optional[str],
+                             stream: 'IO[bytes]') -> str:
+
+        sanitized_filename = secure_filename("unknown.file")
```
I think this would be easier to read if it was in an else block below.

zenmonkeykstop replied:
sure, updated.

legoktm commented on securedrop/journalist_app/col.py:
```diff
@@ -137,15 +138,15 @@ def download_single_file(filesystem_id: str, fn: str) -> werkzeug.Response:
             elif fn.endswith("-doc.gz.gpg") or fn.endswith("doc.zip.gpg"):
-                file = Submission.query.filter(Submission.filename == fn).one()
-                mark_seen([file], journalist)
+                the_file = Submission.query.filter(Submission.filename == fn).one()
```
If we're renaming this maybe a better name would be submitted_file?

This whole block could be more DRY by not duplicating the `mark_seen` call for each if stanza...but that can be done separately.

zenmonkeykstop replied:
:+1:  updated (the name, not the mark_seen() calls)

legoktm commented on securedrop/source_app/main.py:
```diff
@@ -82,7 +84,7 @@ def create() -> werkzeug.Response:
                 create_source_user(
                     db_session=db.session,
                     source_passphrase=codename,
```
This is also documented as taking a DicewarePassphrase, so I don't understand why mypy is happy with it but below wants you to wrap it. In any case, shouldn't the codename variable be turned into a DicewarePassphrase as soon as its pulled out of the session object?

legoktm reviewed (commented):
As I mentioned earlier, I think this could be split a bit to make easier to review:
* datetime changes: utcnow() -> now(utc)
* Introduction of and switch to Storage.get_default()
* Introduction of and switch to InstanceConfig.get_default()

zenmonkeykstop committed:
Reordered requirements installation in Docker env.

Originally, test requirements were installed in the Docker environment after
application requirements. Some dependencies are duplicated in both, with older
versions present in the test requirements. The order of installation was switched
to ensure application requirements were always present and not overridden.

zenmonkeykstop committed:
Updated application dependencies to use Flask 2.0.2

 - click from 6.7 to 8.0.3
 - flask-babel from 011.2 to 2.0.0
 - flask-wtf from 0.14.2 to 1.0.0
 - Flask from 1.0.2 to 2.0.2
 - itsdangerous from 0.24 to 2.0.1
 - jinja2 from 2.11.3 to 3.0.2
 - markupsafe from 1.1.1 to 2.0.1
 - redis from 3.3.6 to 3.5.3
 - rq from 1.1.0 to 1.10.0
 - werkzeug from 0.16.0 to 2.0.2

zenmonkeykstop committed:
Added typing fixes

 - disabled broken typechecking for csrf error handlers
 - disabled typechecking for blueprint 404 error handling workaround
 - updated save_session() with expected type
 - updated _secure_file_stream args to match those expected by Werkzeug 2.0
 - fixed to_json() return type
 - updated file submissions to use expected types and handle case with missing filename
 - ignore babel_instance attribute dynamically added by flask-babel
 - explicitly cast source passphrase for login
 - use expected SessionMixin type instead of Dict[Any, Any] for sessions
 - use BytesIO instead of BufferedBaseIO
 - changed invalid variable name 'file' to 'the_file'

zenmonkeykstop committed:
Updated Flask version in testinfra test variables

zenmonkeykstop committed:
Added app_storage test fixture, updated tests to use it

Some test changes were needed to account for the use of a global Storage object. An app_storage fixture was added, returning a function-scoped Storage object. For tests which access storage directly, that fixture was added to their args and references ito current_app.storage were changed to app_storage. For tests that access storage indirectly (for example, via a test journalist or source application), the Storage.get_default() method must be mocked to return the app_storage fixture. This patch has been added in the source_app and journalist_app fixtures, so it will be applied automatically in tests that use them.

zenmonkeykstop commented:
> As I mentioned earlier, I think this could be split a bit to make easier to review:
> 
>     * datetime changes: utcnow() -> now(utc) 
>     * Introduction of and switch to Storage.get_default()
>     * Introduction of and switch to InstanceConfig.get_default()

This makes sense - I've reordered and merged commits to simplify things and group related changes. Let me know if it's helpful!

zenmonkeykstop committed:
Added option to refresh the global InstanceConfig.

If the instance config is updated in the Journalist Interface, the change should
be reflected in both web applications immediately. To allow for this, an
optional refresh argument forcing a database lookup was added, and the source and
journalist apps were updated to use the refresh option before each request.
Subsequent uses of the InstanceConfig within a request should not need to hit the
database.

zenmonkeykstop committed:
Readability fixes based on PR review.

cfm commented on securedrop/requirements/python3/securedrop-app-code-requirements.txt:
```diff
@@ -108,33 +108,36 @@ cryptography==3.4.7 \
-flask-babel==0.11.2 \
-    --hash=sha256:462a3c599b0ccf426ca1757cc612f1db383844efd346d14170da04c8c76dd521 \
-    --hash=sha256:c0d75710bd4b0fe866f9f2347de6e19208712f9cec006436b4c1c15d4cb0c939
+flask-babel==2.0.0 \
+    --hash=sha256:e6820a052a8d344e178cdd36dd4bb8aea09b4bda3d5f9fa9f008df2c7f2f5468 \
+    --hash=sha256:f9faf45cdb2e1a32ea2ec14403587d4295108f35017a7821a2b1acb8cfd9257d
     # via -r requirements/python3/securedrop-app-code-requirements.in
```
Diff reviewed in <https://github.com/freedomofpress/securedrop-debian-packaging/wiki/Flask-Babel-0.11.2-to-2.0.0>.

cfm commented on securedrop/journalist_app/__init__.py:
```diff
@@ -131,24 +131,25 @@ def setup_g() -> 'Optional[Response]':
-        if app.instance_config.organization_name:
-            g.organization_name = app.instance_config.organization_name
+        if InstanceConfig.get_default().organization_name:
+            g.organization_name = \
+                InstanceConfig.get_default().organization_name  # pylint: disable=assigning-non-slot
         else:
-            g.organization_name = gettext('SecureDrop')
+            g.organization_name = gettext('SecureDrop')  # pylint: disable=assigning-non-slot
```
Nit in passing:  I wondered if this could be simplified, since [`InstanceConfig.organization_name` has `default="SecureDrop"`](https://github.com/freedomofpress/securedrop/blob/develop/securedrop/models.py#L872), but it looks like that default's not applied in the [upgrade migration](https://github.com/freedomofpress/securedrop/blob/f768a1a5aa37e790077dfd38d6e7a0299a4b3a0f/securedrop/alembic/versions/92fba0be98e9_added_organization_name_field_in_.py#L19-L24).

legoktm commented on securedrop/requirements/python3/securedrop-app-code-requirements.txt:
```diff
@@ -226,20 +265,22 @@ python-editor==1.0.3 \
-redis==3.3.6 \
-    --hash=sha256:45682ecf226c7611efe731974c4fa3390170ba045b9cdb26f0051114a5c2a68b \
-    --hash=sha256:f2609a85e5f37f489ba3b5652e1175dc3711c4d7a7818c4f657615810afd23df
+redis==3.5.3 \
```
Reviewed: https://github.com/freedomofpress/securedrop-debian-packaging/wiki/redis-3.3.6-to-3.5.3

conorsch commented:
Diff reviews complete for `click` and `itsdangerous`. Regarding the latter, we've got an upcoming deprecation tracked in https://github.com/freedomofpress/securedrop/issues/6224. For the purposes of this PR, see the comments in https://github.com/pallets/itsdangerous/issues/257 about token length; I have not tried to reproduce what's reported there, but given the scope of these changes, makes sense to evaluate as part of review here.

conorsch reviewed (dismissed):
Looks like `securedrop/requirements/python3/test-requirements.txt` still needs to be updated; it contains `itsdangerous==0.24` (the old version) and also Flash v1 still.

zenmonkeykstop commented:
> Looks like `securedrop/requirements/python3/test-requirements.txt` still needs to be updated; it contains `itsdangerous==0.24` (the old version) and also Flash v1 still.

Yup, that's an outstanding task in the TODOs - current workaround also described there.

legoktm commented:
jinja notes:
* evalcontextfilter is deprecated, there's a new sample nl2br snippet: https://jinja.palletsprojects.com/en/3.0.x/api/?highlight=nl2br#custom-filters
* The with_ and autoescape extensions are deprecated (now always built-in AIUI), so maybe they can be removed from babel.cfg?
* It seems Markup should be imported from markupsafe rather than jinja now
* Do we have a regression test that verifies jinja's autoescaping is turned on? Like `assert Template("{{foo}}").render(foo="<script>") == "&lt;script>"` or something.

I reviewed all the doc and src/ changes, ended up skipping the tests/ changes due to a lack of time.

legoktm commented on securedrop/requirements/python3/securedrop-app-code-requirements.txt:
```diff
@@ -108,33 +108,36 @@ cryptography==3.4.7 \
+    # via
+    #   flask
+    #   flask-wtf
+jinja2==3.0.2 \
```
Reviewed: https://github.com/freedomofpress/securedrop-debian-packaging/wiki/Jinja2-2.11.3-to-3.0.2

nabla-c0d3 commented on securedrop/models.py:
```diff
@@ -877,6 +879,13 @@ def copy(self) -> "InstanceConfig":
         return new
 
+    @classmethod
+    def get_default(cls, refresh: Optional[bool] = False) -> "InstanceConfig":
```
I would put `refresh: bool = False` ; not sure there's a need for refresh to ever be None?

zenmonkeykstop replied:
:heavy_check_mark: - updated.

nabla-c0d3 commented on securedrop/tests/test_store.py:
```diff
@@ -32,30 +32,29 @@ def create_file_in_source_dir(config, filesystem_id, filename):
     return source_directory, file_path
```
To further simplify the tests in this file, I would recommend:

- Removing all mentions of the journalist_app, as it is out of scope for what is being tested here (only the Storage class is being tested; not the full app).
- Not using the storage fixture as I would argue that creating a Storage class should be done in the test code as that's what's being tested: it's not "setup" code. The benefit will be that the tests won't be coupled with the config fixture (which is used in a lot of tests and seems unnecessary here) and the tests will be easier to read/follow.

You can see an example of tis approach above in TestPassphrasesGenerator.

zenmonkeykstop replied:
The `journalist_app` fixture is unused in the majority of the tests alright - there are a few that do require an app context - eg. the shredder tests, but they could probably be moved to a separate file.

I'm less convinced about removing the `app_storage` fixture in all cases. Some of the tests that use it and the `config` fixture are essentially verifying that files are being created in the store in the expected location based on config parameters, and for those it would seem either tautological to just remove `config`, or duplicative to effectively recreate both fixtures in the test. But it can definitely be removed from tests not using `config` and replaced with a test-specific Storage object.

zenmonkeykstop replied:
Actually on further reflection there are a lot more cases where the latter makes sense...

nabla-c0d3 commented on securedrop/tests/test_encryption.py:
```diff
@@ -17,13 +17,14 @@ def test_get_default(self, config):
                 db_session=db.session,
                 source_passphrase=PassphraseGenerator.get_default().generate_passphrase(),
-                source_app_storage=source_app.storage,
+                source_app_storage=app_storage,
```
Not super important, but you might now be able to remove the source_app fixture from this test and the one below 🙌

zenmonkeykstop replied:
Looks like yes to the one below, but no to this one without independently setting up the db and creating an app and passing its context within the test.

zenmonkeykstop committed:
Simplified InstanceConfig constructor as per PR review

zenmonkeykstop committed:
removed unnecessary journalist_app and config fixtures in store tests as per PR review

zenmonkeykstop committed:
removed superfluous source_app fixture from test as per PR review

nabla-c0d3 commented on securedrop/store.py:
```diff
@@ -314,9 +329,13 @@ def save_file_submission(self,
+        if filename is not None:
+            sanitized_filename = secure_filename(filename)
+        else:
+            sanitized_filename = secure_filename("unknown.file")
```
(Not very important) You could simplify this by switching the argument's declaration to:
```
filename: str = "unknown.file",
```
Then no if/else would be needed.

zenmonkeykstop replied:
I kindof like it as is, I think @legoktm's previous point about readability is a good one.

eaon commented on securedrop/requirements/python3/securedrop-app-code-requirements.txt:
```diff
@@ -226,20 +265,22 @@ python-editor==1.0.3 \
-rq==1.1.0 \
-    --hash=sha256:2798d26a7b850e759f23f69695a389d676a9c08f2c14f96f0d34d9648c9d5616 \
-    --hash=sha256:4f27c6a690d1bd02b9157e615d8819555b9b359c0c4ec8ff0013d160c31b40bb
+rq==1.10.0 \
+    --hash=sha256:92950a3e60863de48dd1800882939bbaf089a37497ebf9f2ecf7c9fd0a4c4a95 \
+    --hash=sha256:be09ec43fae9a75a4d26ea3cd520e5fa3ea2ea8cf481be33e6ec9416f0369cac
```
Reviewed: https://github.com/freedomofpress/securedrop-debian-packaging/wiki/rq-1.1.0-to-1.10.0

legoktm commented:
I got through the first half of the test plan today, should finish the rest on Monday.

conorsch reviewed (approved):
Changes look great. Try as I might, I'm unable to break it, which is grand. =) Tested via the prod VM upgrade scenario, and ran through the full test plan. Diff is quite readable, and post-reorg, the code is a bit more intuitive IMO. Great work on this one, @zenmonkeykstop—and thanks also to @legoktm for detailed review.
